File: app/src/main/java/com/pr0gramm/app/services/GifDrawableLoader.java
package com.pr0gramm.app.services;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.net.Uri;
import android.os.Build;

import com.pr0gramm.app.io.Cache;
import com.pr0gramm.app.services.proxy.ProxyService;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.RandomAccessFile;

import javax.inject.Inject;
import javax.inject.Singleton;

import pl.droidsonroids.gif.GifDrawable;
import rx.Emitter;
import rx.Observable;
import rx.Subscription;
import rx.subscriptions.Subscriptions;

import static com.pr0gramm.app.util.AndroidUtility.toFile;
import static java.lang.System.identityHashCode;

/**
 */
@Singleton
public class GifDrawableLoader {
    private static final Logger logger = LoggerFactory.getLogger("GifLoader");

    private final Cache cache;
    private final File fileCache;

    @Inject
    public GifDrawableLoader(Context context, Cache cache, ProxyService proxyService) {
        this.cache = cache;
        this.fileCache = context.getCacheDir();
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    public Observable<DownloadStatus> load(Uri uri) {
        return Observable.<DownloadStatus>create(emitter -> {
            try {
                if ("file".equals(uri.getScheme())) {
                    File file = toFile(uri);
                    emitter.onNext(new DownloadStatus(new GifDrawable(file)));
                    emitter.onCompleted();
                    return;
                }

                try (Cache.Entry entry = cache.get(uri)) {
                    loadGifUsingTempFile(emitter, entry);
                }
            } catch (Throwable error) {
                logger.warn("Error during gif-loading", error);
                emitter.onError(error);
            }
        }, Emitter.BackpressureMode.LATEST).serialize();
    }

    /**
     * Loads the data of the gif into a temporary file. The method then
     * loads the gif from this temporary file. The temporary file is removed
     * after loading the gif (or on failure).
     */
    @SuppressLint("NewApi")
    private void loadGifUsingTempFile(Emitter<? super DownloadStatus> subscriber,
                                      Cache.Entry entry) throws IOException {

        File temporary = new File(fileCache, "tmp" + identityHashCode(subscriber) + ".gif");

        logger.info("storing data into temporary file");

        Subscription subscription = Subscriptions.empty();
        subscriber.setSubscription(subscription);

        try (RandomAccessFile storage = new RandomAccessFile(temporary, "rw")) {
            // remove entry from filesystem now - the system will remove the data
            // when the stream closes.
            //noinspection ResultOfMethodCallIgnored
            temporary.delete();

            // copy data to the file.
            long lastStatusTime = System.currentTimeMillis();
            float contentLength = (float) entry.totalSize();
            try (InputStream stream = entry.inputStreamAt(0)) {
                int length, count = 0;
                byte[] buffer = new byte[16 * 1024];
                while ((length = stream.read(buffer)) >= 0) {
                    storage.write(buffer, 0, length);
                    count += length;

                    if (subscription.isUnsubscribed()) {
                        logger.info("Stopped because the subscriber unsubscribed");
                        return;
                    }

                    // publish download progress every 100ms
                    long now = System.currentTimeMillis();
                    if (now - lastStatusTime > 250) {
                        subscriber.onNext(new DownloadStatus(count / contentLength));
                        lastStatusTime = now;
                    }
                }
            }

            if (subscription.isUnsubscribed()) {
                return;
            }

            try {
                GifDrawable drawable = new GifDrawable(storage.getFD());

                // closing is now delegated to the drawable.
                subscriber.onNext(new DownloadStatus(drawable));
                subscriber.onCompleted();
            } catch (Throwable error) {
                subscriber.onError(error);
            }
        }
    }

    public static class DownloadStatus {
        public final GifDrawable drawable;
        public final float progress;

        DownloadStatus(float progress) {
            this.drawable = null;
            this.progress = progress;
        }

        DownloadStatus(GifDrawable drawable) {
            this.drawable = drawable;
            this.progress = 1.f;
        }

        public boolean finished() {
            return drawable != null;
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/services/GifDrawableLoader.kt
package com.pr0gramm.app.services

import android.net.Uri
import android.os.ParcelFileDescriptor
import androidx.core.net.toFile
import com.pr0gramm.app.Duration.Companion.millis
import com.pr0gramm.app.Logger
import com.pr0gramm.app.io.Cache
import com.pr0gramm.app.util.debugOnly
import com.pr0gramm.app.util.isLocalFile
import com.pr0gramm.app.util.readStream
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import pl.droidsonroids.gif.GifAnimationMetaData
import pl.droidsonroids.gif.GifDrawable
import pl.droidsonroids.gif.GifDrawableBuilder
import pl.droidsonroids.gif.GifOptions
import java.io.File
import java.io.RandomAccessFile

/**
 */
class GifDrawableLoader(private val cache: Cache) {
    private val logger = Logger("GifLoader")

    fun load(uri: Uri): Flow<State> {
        val result = flow {
            try {
                if (uri.isLocalFile) {
                    emit(State(createGifDrawable(uri.toFile())))
                } else {
                    emitAll(loadGifUsingCache(uri))
                }

            } catch (error: Throwable) {
                logger.warn(error) { "Error during gif-loading" }
                throw error
            }
        }

        return result.flowOn(Dispatchers.IO)
    }

    /**
     * Loads the data of the gif into a temporary file. The method then
     * loads the gif from this temporary file. The temporary file is removed
     * after loading the gif (or on failure).
     */
    private fun loadGifUsingCache(uri: Uri): Flow<State> {
        return flow {
            logger.info { "Getting file reference from cached file" }

            cache.get(uri).use { entry ->
                if (entry.fractionCached < 1.0) {
                    // read the gif once so it is in the cache.
                    emitAll(readToCache(entry))
                }

                // the file should now be fully cached, we should be able to read it now.
                entry.file?.let { file ->
                    // we should now have a file there.
                    return@flow emit(State(createGifDrawable(file)))
                }
            }

            // crap, no file in cache, we cant do much now.
            logger.error { "No file in cache after reading the full file." }
        }
    }

    private fun readToCache(entry: Cache.Entry): Flow<State> {
        return flow {
            // read file once so we have it in the cache
            val publishState = OnceEvery(millis(100))
            val contentLength = entry.totalSize.toFloat()
            entry.inputStreamAt(0).use { stream ->
                var count = 0

                readStream(stream) { _, length ->
                    currentCoroutineContext().ensureActive()

                    debugOnly {
                        // simulate slow network connection in debug mode
                        Thread.sleep(if (Math.random() < 0.1) 100 else 0)
                    }

                    count += length

                    publishState {
                        emit(State(progress = count / contentLength))
                    }
                }
            }
        }
    }

    private fun createGifDrawable(file: File): GifDrawable {
        RandomAccessFile(file, "r").use { storage ->
            val meta = ParcelFileDescriptor.dup(storage.fd).use {
                GifAnimationMetaData(it.fileDescriptor)
            }

            val sampleSize = intArrayOf(1, 2, 3, 4, 5).firstOrNull { meta.width / it <= 1024 } ?: 6
            logger.info { "Loading gif ${meta.width}x${meta.height} with sampleSize $sampleSize" }

            return GifDrawableBuilder()
                .from(storage.fd)
                .options(GifOptions().apply {
                    setInSampleSize(sampleSize)
                    setInIsOpaque(true)
                })
                .renderingTriggeredOnDraw(true)
                .build()
        }
    }

    class State(
        val drawable: GifDrawable? = null,
        val progress: Float = 1.0f
    )
}
